.github/copilot-instructions.md

```diff
@@ -0,0 +1,109 @@
+  - Type checking: MyPy
+  - Testing: pytest with plain functions and fixtures
+- Inline code documentation:
+  - File headers should be short and concise:
+    ```python
+    """Integration for Peblar EV chargers."""
+    ```
+  - Every method and function needs a docstring:
+    ```python
+    async def async_setup_entry(hass: HomeAssistant, entry: PeblarConfigEntry) -> bool:
+        """Set up Peblar from a config entry."""
+        ...
+    ```
+- All code and comments and other text are written in American English
+- Follow existing code style patterns as much as possible
+- Core locations:
+  - Shared constants: `homeassistant/const.py`, use them instead of hardcoding
+    strings or creating duplicate integration constants.
+  - Integration files:
+    - Constants: `homeassistant/components/{domain}/const.py`
+    - Models: `homeassistant/components/{domain}/models.py`
+    - Coordinator: `homeassistant/components/{domain}/coordinator.py`
+    - Config flow: `homeassistant/components/{domain}/config_flow.py`
+    - Platform code: `homeassistant/components/{domain}/{platform}.py`
+- All external I/O operations must be async
+- Async patterns:
+  - Avoid sleeping in loops
+  - Avoid awaiting in loops, gather instead
+  - No blocking calls
+- Polling:
+  - Follow update coordinator pattern, when possible
+  - Polling interval may not be configurable by the user
+  - For local network polling, the minimum interval is 5 seconds
+  - For cloud polling, the minimum interval is 60 seconds
+- Error handling:
+  - Use specific exceptions from `homeassistant.exceptions`
+  - Setup failures:
+    - Temporary: Raise `ConfigEntryNotReady`
+    - Permanent: Use `ConfigEntryError`
+  - Raised error should support translations:
+    ```python
+    except PeblarConnectionError as error:
+        raise HomeAssistantError(
+            translation_domain=DOMAIN,
+            translation_key="communication_error",
+            translation_placeholders={"error": str(error)},
+        ) from error
+    ```
+- Logging:
+  - Message format:
+    - No periods at end
+    - No integration names or domains (added automatically)
+    - No sensitive data (keys, tokens, passwords), if when those are incorrect.
+  - Be very restrictive on the use if logging info messages, use debug for
+    anything which is not targeting the user.
+  - Use lazy logging (no f-strings):
+    ```python
+    _LOGGER.debug("This is a log message with %s", variable)
+    ```
+- Entities:
+  - Ensure unique IDs for state persistence:
+    - Unique IDs should not contain values that are subject to user change.
+    - An ID needs to be unique per platform, not per integration.
+    - The ID does not have to contain the integration domain or platform.
+    - Acceptable examples:
+      - Serial number of a device
+      - MAC address of a device formatted using `homeassistant.helpers.device_registry.format_mac`
+        Do not obtain the MAC address thru arp cache of local network access,
```

Comment:
Out of curiosity, why?

Reply:
Because it ain't reliable. You could be returned the mac of a switch in between instead of the actual end-device.

Reply:
Is that done using getmac?

Reply:
Yup.